pubsub/acks_test.go

```diff
@@ -342,3 +344,88 @@ func TestReceiveReturnsErrorFromSendAcks(t *testing.T) {
 		}
 	}
 }
+
+// callbackDriverSub implements driver.Subscription and allows something like
+// monkey patching of both its ReceiveBatch and SendAcks methods.
+type callbackDriverSub struct {
+	driver.Subscription
+	receiveBatch func(context.Context) ([]*driver.Message, error)
+	sendAcks     func(context.Context, []driver.AckID) error
+}
+
+func (s *callbackDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	return s.receiveBatch(ctx)
+}
+
+func (s *callbackDriverSub) SendAcks(ctx context.Context, acks []driver.AckID) error {
+	return s.sendAcks(ctx, acks)
+}
+
+func (s *callbackDriverSub) IsRetryable(error) bool { return false }
+
+func (s *callbackDriverSub) ErrorCode(error) gcerrors.ErrorCode { return gcerrors.Internal }
+
+// This test detects the root cause of
+// https://github.com/google/go-cloud/issues/1238.
+// If the issue is present, this test times out. The problem was that when
+// there were no messages available from the provider,
+// pubsub.Subscription.Receive would spin trying to get more messages without
+// checking to see if an unrecoverable error had occurred while sending a batch
+// of acks to the provider.
+func TestReceiveReturnsAckErrorOnNoMoreMessages(t *testing.T) {
+	// If SendAcks fails, the error is returned via receive.
+	ctx := context.Background()
+	serr := errors.New("unrecoverable error")
+	receiveHappened := make(chan struct{})
+	ackHappened := make(chan struct{})
+	ds := &callbackDriverSub{
+		// First call to receiveBatch will return a single message.
+		receiveBatch: func(context.Context) ([]*driver.Message, error) {
+			ms := []*driver.Message{{AckID: 1}}
+			return ms, nil
+		},
+		sendAcks: func(context.Context, []driver.AckID) error {
+			ackHappened <- struct{}{}
+			return serr
+		},
+	}
+	sub := pubsub.NewSubscription(ds, nil)
+	defer sub.Shutdown(ctx)
+	m, err := sub.Receive(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	m.Ack()
+
+	// Second call to receiveBatch will wait for the pull from the
+	// receiveHappened channel below, and return a nil slice of messages.
+	ds.receiveBatch = func(context.Context) ([]*driver.Message, error) {
```

Comment:
`ReceiveBatch` is documented as needing to be safe to call from concurrent goroutines. Assigning to `ds.receiveBatch` without synchronization is a data race IIUC.

Reply:
I agree that there could be a data race in the case where multiple goroutines simultaneously call `sub.Receive`, but in this test only one call to `sub.Receive` ever occurs. Our CI setup runs the tests with -race, and that should detect this situation if it ever comes up for some reason.

Reply:
I agree with Issac—in the context of this test, this is ok.

Reply:
Pardon me. I realize now that my reply didn't really address your concern. Have to ponder this for a minute...

Reply:
I can see how someone looking at the code could worry that the first call to `sub.Receive` could spawn a goroutine that would indirectly call `ds.receiveBatch`. Even if it doesn't now, it could later. I added a mutex so no one has to worry about it or spelunk the code to find out.

Reply:
Thanks @ijt. That's more what I was worried about: even if it isn't a race now, a subtle change to how Pub/Sub operates could render the test racy.